Source/WebCore/style/ElementRuleCollector.cpp

```diff
@@ -362,15 +378,31 @@ void ElementRuleCollector::collectMatchingShadowPseudoElementRules(const MatchRe
         collectMatchingRulesForList(rules.shadowPseudoElementRules(pseudoId), matchRequest);
 }
 
+static RuleSet* getUserRules(const ScopeRuleSets& ruleSets)
+{
+    return ruleSets.userStyle();
+}
+
 void ElementRuleCollector::matchUserRules()
 {
-    if (!m_userStyle)
-        return;
-    
     clearMatchedRules();
 
-    MatchRequest matchRequest(*m_userStyle);
-    collectMatchingRules(matchRequest);
+    if (m_userStyle) {
+        MatchRequest matchRequest(*m_userStyle);
+        collectMatchingRules(matchRequest);
+    }
+
+    auto* parent = element().parentElement();
+    if (parent && parent->shadowRoot())
+        matchSlottedPseudoElementRules(getUserRules);
+
+    if (element().shadowRoot())
+        matchHostPseudoClassRules(getUserRules);
+
+    if (element().isInShadowTree()) {
+        matchShadowPseudoElementRules(getUserRules);
+        matchPartPseudoElementRules(getUserRules);
+    }
```

Comment:
Some of this can probably be factored into a shared function with author style matching.

Reply:
we definitely could!  tho it would likely involve creating a `collectMatchingRules(const GetRulesFunction& getRules)` for the same reasons as above, so not sure how you feel about that 😅